utils/core/src/serde/byte_reader.rs

```diff
@@ -0,0 +1,166 @@
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u16 value could not be read from `self`.
+    fn read_u16(&mut self) -> Result<u16, DeserializationError> {
+        let bytes = self.read_array::<2>()?;
+        Ok(u16::from_le_bytes(bytes))
+    }
+
+    /// Returns a u32 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u32 value could not be read from `self`.
+    fn read_u32(&mut self) -> Result<u32, DeserializationError> {
+        let bytes = self.read_array::<4>()?;
+        Ok(u32::from_le_bytes(bytes))
+    }
+
+    /// Returns a u64 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u64 value could not be read from `self`.
+    fn read_u64(&mut self) -> Result<u64, DeserializationError> {
+        let bytes = self.read_array::<8>()?;
+        Ok(u64::from_le_bytes(bytes))
+    }
+
+    /// Returns a u128 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u128 value could not be read from `self`.
+    fn read_u128(&mut self) -> Result<u128, DeserializationError> {
+        let bytes = self.read_array::<16>()?;
+        Ok(u128::from_le_bytes(bytes))
+    }
+
+    /// Reads a deserializable value from `self`.
+    ///
+    /// # Panics
+    /// Panics if the value could not be read from `self`.
+    fn read<D: Deserializable>(&mut self) -> Result<D, DeserializationError>
+    where
+        Self: Sized,
+    {
+        D::read_from(self)
+    }
+}
+
+// SLICE READER
+// ================================================================================================
+
+/// Implements [ByteReader] trait for a slice of bytes.
+pub struct SliceReader<'a> {
+    source: &'a [u8],
+    pos: usize,
+}
+
+impl<'a> SliceReader<'a> {
+    /// Creates a new slice reader from the specified slice.
+    pub fn new(source: &'a [u8]) -> Self {
+        SliceReader { source, pos: 0 }
+    }
+}
+
+impl<'a> ByteReader for SliceReader<'a> {
+    fn read_u8(&mut self) -> Result<u8, DeserializationError> {
+        self.check_eor(1)?;
+        let result = self.source[self.pos];
+        self.pos += 1;
+        Ok(result)
+    }
+
+    fn peek_u8(&self) -> Result<u8, DeserializationError> {
+        self.check_eor(1)?;
+        Ok(self.source[self.pos])
+    }
+
+    fn read_vec(&mut self, len: usize) -> Result<Vec<u8>, DeserializationError> {
+        self.check_eor(len)?;
+        let result = self.source[self.pos..self.pos + len].to_vec();
+        self.pos += len;
+        Ok(result)
+    }
+
+    fn read_array<const N: usize>(&mut self) -> Result<[u8; N], DeserializationError> {
+        self.check_eor(N)?;
+        let bytes = &self.source[self.pos..self.pos + N];
+        self.pos += N;
+        Ok(bytes.try_into().expect("slice to array conversion failed"))
```

Comment:
Using a `try_into` might be slightly more expensive than it needs to be because the compiler will have to worry to display the expected error message: `slice to array conversion failed"`

This is in fact infallible, so we probably shouldn't add more cycles for something that will never happen (as the length of the bytes is already checked)

We could, instead

```rust
self.check_eor(N)?;
let mut buf = [0_u8; N];
buf.copy_from_slice(&self.source[self.pos..self.pos + N]);
self.pos += N;
Ok(buf)
```

Reply:
Good idea! Updated.